### PR TITLE
refactor: change the schema type of DataFrame to PortexRecordBase

### PR DESCRIPTION
**Pull request by Lee-000**

### CODE REVIEW

**Comment by GravitiRob** [comment]
Task linked: [CU-29hwhv4 SDK 将 record 类型 + 类 record 的模板类型与其他类型区分](https://app.clickup.com/t/29hwhv4)

**Review by graczhual — APPROVED** [review]

File: graviti/manager/sheets.py [hunk 1]
```diff
@@ -63,7 +63,7 @@ def _init_data(self) -> None:
             )
             paging_lists = factory.create_lists(schema.get_keys())
             self._data[sheet_name] = DataFrame._from_paging(  # pylint: disable=protected-access
-                paging_lists, schema
+                paging_lists, schema  # type: ignore[arg-type]
```
Comment (linjiX) [inline graviti/manager/sheets.py:66]: where these `ignore`s come from?

File: graviti/dataframe/frame.py [hunk 2]
```diff
@@ -637,7 +643,7 @@ def _get_process(value: Any, schema: pt.PortexType) -> Tuple[Callable[[Any], Any
 
     @staticmethod
     def _process_array(
-        values: Iterable[Any], schema: pt.PortexType
+        values: Iterable[Any], schema: RecordLikeType
```
Comment (AChenQ) [inline graviti/dataframe/frame.py:646]: Don't need to change schema here

**Review by AChenQ — CHANGES_REQUESTED** [review]

File: graviti/portex/template.py [hunk 3]
```diff
@@ -90,6 +90,21 @@ def get_keys(self, type_name: Optional[str] = None) -> List[Tuple[str, ...]]:
         return self.to_builtin().get_keys(type_name)
 
 
+class PortexExternalRecordType(PortexExternalType):  # pylint: disable=abstract-method
```
Comment (linjiX) [inline graviti/portex/template.py:93]: can we make a class, maybe `RecordBase` as the base class for both `record` and other record-like types?

File: graviti/manager/sheets.py (hunk 1 above)
Comment (Lee-000) [reply graviti/manager/sheets.py:66]: The type of `schema` input of `Dataframe._from_paging` has been changed to record-like portex type, which is inconsistent with the return type of `PortexType.from_yaml`.

File: graviti/portex/factory.py [hunk 4]
```diff
@@ -30,6 +30,8 @@
 class Factory:
     """The base class of the template factory."""
 
+    _fields_key: str
```
Comment (linjiX) [inline graviti/portex/factory.py:33]: why factory need `_fields_key`?

File: graviti/portex/builder.py [hunk 5]
```diff
@@ -238,18 +239,29 @@ def build(self) -> Type["PortexExternalType"]:
...
-        package[self._name] = type_
+        type_ = self._get_external_type(params, factory)
+        self._builder.package[self._name] = type_
 
         return type_
 
+    def _get_external_type(
+        self, params: Params, factory: Union[TypeFactory, VariableFactory]
+    ) -> Type[PortexExternalType]:
+        parent_types: List[Type["PortexType"]] = [PortexExternalType]
```
Comment (linjiX) [inline graviti/portex/builder.py:250]: `parent_types` -> `bases` ?

File: graviti/portex/builder.py [hunk 6]
```diff
@@ -238,18 +239,29 @@ def build(self) -> Type["PortexExternalType"]:
...
+        class_attrs: Dict[str, Any] = {
+            "params": params,
+            "factory": factory,
+            "package": self._builder.package,
+        }
+
+        if isinstance(factory, TypeFactory):
```
Comment (linjiX) [inline graviti/portex/builder.py:257]: I think factory should always be `TypeFactory` here, under what situation the factory will not be a `TypeFactory`?

**Review by linjiX — CHANGES_REQUESTED** [review]

**Review by linjiX — APPROVED** [review]

**Review by AChenQ — APPROVED** [review]

File: graviti/dataframe/frame.py [hunk 7]
```diff
@@ -69,13 +69,13 @@ class DataFrame(Container):
     has_keys = True
```
Comment (graczhual) [inline graviti/dataframe/frame.py:69]: do we need to modify the type of schema in operations?

**Review by graczhual — APPROVED** [review]

**Review by linjiX — APPROVED** [review]

**Pull request merged**